### PR TITLE
Finish Translating Testing Environments

### PR DESCRIPTION
**Pull request by mohamedsgap**

### CODE REVIEW

**Comment by mohamedsgap** [comment]
@3imed-jaberi  @asantarissy Hey guys, I need your reviews for this pull request, I've just finished translating Testing Environemnts

File: content/docs/testing-environments.md [hunk 1]
```diff
@@ -1,58 +1,60 @@
...
 
-This is especially useful for data fetching. It is usually preferable to use "fake" data for tests to avoid the slowness and flakiness due to fetching from real API endpoints [<small>(example)</small>](/docs/testing-recipes.html#data-fetching). This helps make the tests predictable. Libraries like [Jest](https://jestjs.io/) and [sinon](https://sinonjs.org/), among others, support mocked functions. For end-to-end tests, mocking network can be more difficult, but you might also want to test the real API endpoints in them anyway.
+### محاكاة الدوال {#mocking-functions}
 
-### Mocking modules {#mocking-modules}
+عند كتابة الاختبارات ، نود أن نحاكى  الكود الخاص بنا الذى لا يحتوى على مكافئ له فى بيئة الاختبار الخاصة بنا (على سبيل المثال التحقق من حالة `navigator.onLin` داخل Node.js). يمكن للاختبارات أيضًا التجسس على بعض الوظائف ، ومراقبة كيفية تفاعل أجزاء أخرى من الاختبار معها. من المفيد عندئذ أن تكون قادرًا على محاكاة هذه الدوال باصدارات سهلة الاستخدام بشكل الانتقائى.
```
Comment (3imed-jaberi) [inline content/docs/testing-environments.md]:
```suggestion
عند كتابة الاختبارات ، نود أن نحاكى  الكود الخاص بنا الذى لا يحتوى على  مُمَاثل له فى بيئة الاختبار الخاصة بنا (على سبيل المثال التحقق من حالة `navigator.onLin` داخل Node.js). يمكن للاختبارات أيضًا التجسس على بعض الدوال، ومراقبة كيفية تفاعل أجزاء أخرى من الاختبار معها. من المفيد عندئذ أن تكون قادرًا على محاكاة هذه الدوال باصدارات سهلة الاستخدام بشكل الانتقائى.
```

File: content/docs/testing-environments.md [hunk 2]
```diff
@@ -1,58 +1,60 @@
...
 
-Some components have dependencies for modules that may not work well in test environments, or aren't essential to our tests. It can be useful to selectively mock these modules out with suitable replacements [<small>(example)</small>](/docs/testing-recipes.html#mocking-modules).
+هذا مفيد بشكل خاص لجلب البيانات. من المفضل عادة استخدام البيانات "المزيفة" للاختبارات لتجنب البطء والضعف بسبب جلب نقاط نهاية API الحقيقية [<small>(مثال) </small>](/docs/testing-recipes.html#data-fetching ). هذا يساعد على جعل الاختبارات يمكن التنبؤ بها. المكتبات مثل [Jest](https://jestjs.io/) و [sinon](https://sinonjs.org/)، من بين أمور أخرى ، تدعم محاكاه الدوال. بالنسبة للاختبارات الشاملة ، يمكن أن تكون محاكاة الشبكة أكثر صعوبة ، ولكن قد ترغب أيضًا في اختبار نقاط النهاية الحقيقية لواجهة برمجة التطبيقات فيها.
```
Comment (3imed-jaberi) [inline content/docs/testing-environments.md]:
```suggestion
هذا مفيد بشكل خاص لجلب البيانات. من المفضل عادة استخدام البيانات "المزيفة" للاختبارات لتجنب البطء والضعف بسبب الجلب من نقاط الوصول النهائية لواجهة برمجة التطبيقات الحقيقية. [<small>(مثال) </small>](/docs/testing-recipes.html#data-fetching ). هذا يساعد على جعل الاختبارات يمكن التنبؤ بها. مكتبات مثل [Jest](https://jestjs.io/) و [sinon](https://sinonjs.org/)، من بين أمور أخرى ، تدعم محاكاه الدوال. بالنسبة لاختبارات الـ end-to-end ، يمكن أن تكون محاكاة الشبكة أكثر صعوبة ، ولكن قد ترغب أيضًا في اختبار نقاط الوصول الحقيقية لواجهة برمجة التطبيقات فيها.
```

File: content/docs/testing-environments.md [hunk 3]
```diff
@@ -1,58 +1,60 @@
...
 
-End-to-end tests are useful for testing longer workflows, especially when they're critical to your business (such as payments or signups). For these tests, you'd probably want to test both how a real browser renders the whole app, fetches data from the real API endpoints, uses sessions and cookies, navigates between different links. You might also likely want to make assertions not just on the DOM state, but on the backing data as well (e.g. to verify whether the updates have been persisted to the database).
+في بعض الأحيان ، قد لا ترغب في محاكاة العداد. على سبيل المثال ، ربما تقوم باختبار رسم متحرك ، أو تتفاعل مع نقطة نهاية حساسة للتوقيت (like an API rate limiter). تتيح لك المكتبات التي بها محاكاه العداد تمكينها وتعطيلها على أساس كل اختبار / مجموعة ، بحيث يمكنك اختيار كيفية تشغيل هذه الاختبارات بشكل صريح.
```
Comment (3imed-jaberi) [inline content/docs/testing-environments.md]:
```suggestion
في بعض الأحيان ، قد لا ترغب في محاكاة العداد. على سبيل المثال ، ربما تقوم باختبار رسم متحرك ، أو تتفاعل مع نقطة نهاية حساسة للتوقيت (مثل واجهة برمجة التطبيقات من نوع API rate limiter). تتيح لك المكتبات التي بها محاكاه العداد تمكينها وتعطيلها على أساس كل اختبار / مجموعة ، بحيث يمكنك اختيار كيفية تشغيل هذه الاختبارات بشكل صريح.
```

File: content/docs/testing-environments.md [hunk 4]
```diff
@@ -1,58 +1,60 @@
...
 
-In this scenario, you would use a framework like [Cypress](https://www.cypress.io/) or a library like [puppeteer](https://github.com/GoogleChrome/puppeteer) so you can navigate between multiple routes and assert on side effects not just in the browser, but potentially on the backend as well.
+### اختبارات طرف الى طرف {#end-to-end-tests-aka-e2e-tests}
```
Comment (3imed-jaberi) [inline content/docs/testing-environments.md]:
```suggestion
### اختبارات end-to-end {#end-to-end-tests-aka-e2e-tests}
```

File: content/docs/testing-environments.md [hunk 5]
```diff
@@ -1,58 +1,60 @@
...
+
+تعد اختبارات طرف الى طرف مفيدة لاختبار سير عمل أطول ، خاصةً عندما تكون مهمة لنشاطك التجاري (مثل المدفوعات أو الاشتراكات). بالنسبة لهذه الاختبارات ، قد ترغب في اختبار كلٍّ من كيفية عرض المتصفح الحقيقي للتطبيق بأكمله ، وجلب البيانات من نقاط النهاية الحقيقية لواجهة برمجة التطبيقات ، واستخدام الجلسات وملفات تعريف الارتباط ، والتنقل بين الروابط المختلفة. قد ترغب أيضًا في تقديم تأكيدات ليس فقط في حالة DOM ، ولكن أيضًا على بيانات النسخ الاحتياطي (على سبيل المثال للتحقق من استمرار التحديثات في قاعدة البيانات).
```
Comment (3imed-jaberi) [inline content/docs/testing-environments.md]:
```suggestion
تعد اختبارات end-to-end مفيدة لاختبار سير عمل أطول ، خاصةً عندما تكون مهمة لنشاطك التجاري (مثل المدفوعات أو الاشتراكات). بالنسبة لهذه الاختبارات ، قد ترغب في اختبار كلٍّ من كيفية عرض المتصفح الحقيقي للتطبيق بأكمله ، وجلب البيانات من نقاط الوصول الحقيقية لواجهة برمجة التطبيقات ، واستخدام الجلسات (session) وملفات تعريف الارتباط (cookies)، والتنقل بين الروابط المختلفة. قد ترغب أيضًا في تقديم تأكيدات ليس فقط في حالة DOM ، ولكن أيضًا على بيانات النسخ الاحتياطي (على سبيل المثال للتحقق من استمرار التحديثات في قاعدة البيانات).
```

File: content/docs/testing-environments.md [hunk 6]
```diff
@@ -1,58 +1,60 @@
...
+
+في هذا السيناريو ، يمكنك استخدام إطار عمل مثل [Cypress](https://www.cypress.io/) أو مكتبة مثل [puppeteer](https://github.com/GoogleChrome/puppeteer) حتى تتمكن من التنقل بين طرق متعددة والتأكيد على الآثار الجانبية ليس فقط في المتصفح ، ولكن يحتمل أن يكون على الواجهة الخلفية أيضًا.
```
Comment (3imed-jaberi) [inline content/docs/testing-environments.md]:
```suggestion
في هذا السيناريو ، يمكنك استخدام إطار عمل مثل [Cypress](https://www.cypress.io/) أو مكتبة مثل [puppeteer](https://github.com/GoogleChrome/puppeteer) حتى تتمكن من التنقل و تصفح بين نقاط الوصول والطرق (routes) المتعددة والتأكيد على الآثار الجانبية ليس فقط في المتصفح ، ولكن يحتمل أن يكون على الواجهة الخلفية (back-end) أيضًا.
```

**Comment by mohamedsgap** [comment]
@3imed-jaberi @asantarissy @Aissaoui-Ahmed improved the translation after the requested changes! please, check it out :)

File: content/docs/testing-environments.md [hunk 7]
```diff
@@ -1,58 +1,60 @@
...
+تمامًا كما في المتصفح الحقيقي ، تتيح لنا jsdom تصميم تفاعلات المستخدم ؛ يمكن للاختبارات إرسال الأحداث على عقد DOM ، ثم مراقبة الآثار الجانبية لهذه الإجراءات والتأكيد عليها [<small>(مثال)</small>](/docs/testing-recipes.html#events).
 
-If you're writing a library that tests mostly browser-specific behavior, and requires native browser behavior like layout or real inputs, you could use a framework like [mocha.](https://mochajs.org/)
 
-In an environment where you _can't_ simulate a DOM (e.g. testing React Native components on Node.js), you could use [event simulation helpers](https://reactjs.org/docs/test-utils.html#simulate) to simulate interactions with elements. Alternately, you could use the `fireEvent` helper from [`@testing-library/react-native`](https://testing-library.com/docs/native-testing-library).
+يمكن كتابة جزء كبير من اختبارات واجهة المستخدم باستخدام الإعداد أعلاه: استخدام Jest كمنفذ للاختبار ، يتم تصييره إلى jsdom ، مع تفاعلات المستخدم المحددة كسلسلة من أحداث المتصفح ، مدعومًا بـ `act ()` helper [<small>(مثال)</small>](/docs/testing-recipes.html). على سبيل المثال ، تتم كتابة الكثير من اختبارات React الخاصة مع هذه المجموعة.
```
[the remark this answers is not in the document]
Comment (3imed-jaberi) [reply content/docs/testing-environments.md]: @mohamedsgap !!

File: content/docs/testing-environments.md [hunk 8]
```diff
@@ -1,58 +1,60 @@
...
 
-Frameworks like [Cypress](https://www.cypress.io/), [puppeteer](https://github.com/GoogleChrome/puppeteer) and [webdriver](https://www.seleniumhq.org/projects/webdriver/) are useful for running [end-to-end tests](#end-to-end-tests-aka-e2e-tests).
+إذا كنت تكتب مكتبة تختبر في الغالب السلوك الخاص بالمتصفح ، وتتطلب سلوك المستعرض الأصلي مثل التخطيط أو المدخلات الحقيقية ، يمكنك استخدام إطار عمل مثل [mocha.](https://mochajs.org/)
```
[the remark this answers is not in the document]
Comment (3imed-jaberi) [reply content/docs/testing-environments.md]: @mohamedsgap !!

File: content/docs/testing-environments.md [hunk 9]
```diff
@@ -1,58 +1,60 @@
...
 
-### Mocking functions {#mocking-functions}
+في بيئة لا يمكنك فيها محاكاة DOM (على سبيل المثال ، اختبار مكونات React Native على Node.js) ، يمكنك استخدام [مساعدي محاكاة الأحداث](https://reactjs.org/docs/test-utils.html#simulate) لمحاكاة التفاعلات مع العناصر. بالتناوب ، يمكنك استخدام "fireEvent" المساعد من [`@ testing-library / react-native`](https://testing-library.com/docs/native-testing-library).
```
[the remark this answers is not in the document]
Comment (3imed-jaberi) [reply content/docs/testing-environments.md]: @mohamedsgap !!

File: content/docs/testing-environments.md [hunk 10]
```diff
@@ -1,58 +1,60 @@
...
 
-When writing tests, we'd like to mock out the parts of our code that don't have equivalents inside our testing environment (e.g. checking `navigator.onLine` status inside Node.js). Tests could also spy on some functions, and observe how other parts of the test interact with them. It is then useful to be able to selectively mock these functions with test-friendly versions.
+اطارات العمل مثل [Cypress](https://www.cypress.io/) و [puppeteer](https://github.com/GoogleChrome/puppeteer) و [webdriver](https://www.seleniumhq.org/projects / webdriver /) مفيدة لتشغيل [end-to-end test ](#end-to-end-tests-aka-e2e-tests).
```
[the remark this answers is not in the document]
Comment (3imed-jaberi) [reply content/docs/testing-environments.md]: @mohamedsgap !!

File: content/docs/testing-environments.md (hunk 1 above)
Comment (3imed-jaberi) [reply content/docs/testing-environments.md]: @mohamedsgap !!

File: content/docs/testing-environments.md [hunk 11]
```diff
@@ -1,58 +1,60 @@
...
+تعمل الاختبارات غالبًا في بيئة دون الوصول إلى سطح التصيير الحقيقي مثل المستعرض. بالنسبة لهذه البيئات ، نوصي بمحاكاة مستعرض باستخدام [`jsdom`](https://github.com/jsdom/jsdom) ، وهو تطبيق متصفح خفيف الوزن يعمل داخل Node.js.
 
-Just like in a real browser, jsdom lets us model user interactions; tests can dispatch events on DOM nodes, and then observe and assert on the side effects of these actions [<small>(example)</small>](/docs/testing-recipes.html#events).
+في معظم الحالات ، يتصرف jsdom كالمتصفح العادي ، لكن ليس به ميزات مثل [layout and navigation ](https://github.com/jsdom/jsdom#unimplemented-parts-of-the-web-platform). لا يزال هذا مفيدًا لمعظم اختبارات المكونات المستندة إلى الويب ، لأنه يعمل بشكل أسرع من الاضطرار إلى بدء تشغيل مستعرض لكل اختبار. يتم تشغيله أيضًا في نفس عملية الاختبارات الخاصة بك ، حتى تتمكن من كتابة التعليمات البرمجية لفحصها وتأكيدها على DOM.
```
[the remark this answers is not in the document]
Comment (3imed-jaberi) [reply content/docs/testing-environments.md]: @mohamedsgap !!

File: content/docs/testing-environments.md [hunk 12]
```diff
@@ -1,58 +1,60 @@
...
+- تُستخدم الاختبارات من طرف إلى طرف لاختبار تدفقات أطول عبر عدة صفحات ، وتتطلب [إعداد مختلف](#end-to-end-tests-aka-e2e-tests).
 
-Tests often run in an environment without access to a real rendering surface like a browser. For these environments, we recommend simulating a browser with [`jsdom`](https://github.com/jsdom/jsdom), a lightweight browser implementation that runs inside Node.js.
+### محاكاة سطح التصيير {#mocking-a-rendering-surface}
 
-In most cases, jsdom behaves like a regular browser would, but doesn't have features like [layout and navigation](https://github.com/jsdom/jsdom#unimplemented-parts-of-the-web-platform). This is still useful for most web-based component tests, since it runs quicker than having to start up a browser for each test. It also runs in the same process as your tests, so you can write code to examine and assert on the rendered DOM.
+تعمل الاختبارات غالبًا في بيئة دون الوصول إلى سطح التصيير الحقيقي مثل المستعرض. بالنسبة لهذه البيئات ، نوصي بمحاكاة مستعرض باستخدام [`jsdom`](https://github.com/jsdom/jsdom) ، وهو تطبيق متصفح خفيف الوزن يعمل داخل Node.js.
```
[the remark this answers is not in the document]
Comment (3imed-jaberi) [reply content/docs/testing-environments.md]: @mohamedsgap !!

File: content/docs/testing-environments.md [hunk 13]
```diff
@@ -1,58 +1,59 @@
...
 permalink: docs/testing-environments.html
 prev: testing-recipes.html
 ---
 
 <!-- This document is intended for folks who are comfortable with JavaScript, and have probably written tests with it. It acts as a reference for the differences in testing environments for React components, and how those differences affect the tests that they write. This document also assumes a slant towards web-based react-dom components, but has notes for other renderers. -->
 
-This document goes through the factors that can affect your environment and recommendations for some scenarios.
+يتناول هذا المستند العوامل التي يمكن أن تؤثر على بيئتك والتوصيات المتعلقة ببعض السيناريوهات.
 
-### Test runners {#test-runners}
```
Comment (Aissaoui-Ahmed) [inline content/docs/testing-environments.md]: Remove line 12

File: content/docs/testing-environments.md [hunk 14]
```diff
@@ -1,58 +1,59 @@
...
+في معظم الحالات ، يتصرف jsdom كالمتصفح العادي ، لكن ليس به ميزات مثل [layout and navigation ](https://github.com/jsdom/jsdom#unimplemented-parts-of-the-web-platform). لا يزال هذا مفيدًا لمعظم اختبارات المكونات المستندة إلى الويب ، لأنه يعمل بشكل أسرع من الاضطرار إلى بدء تشغيل مستعرض لكل اختبار. يتم تشغيله أيضًا في نفس عملية الاختبارات الخاصة بك ، حتى تتمكن من كتابة التعليمات البرمجية لفحصها وتأكيدها على DOM.
 
-Just like in a real browser, jsdom lets us model user interactions; tests can dispatch events on DOM nodes, and then observe and assert on the side effects of these actions [<small>(example)</small>](/docs/testing-recipes.html#events).
+تمامًا كما في المتصفح الحقيقي ، تتيح لنا jsdom تصميم تفاعلات المستخدم ؛ يمكن للاختبارات إرسال الأحداث على عقد DOM ، ثم مراقبة الآثار الجانبية لهذه الإجراءات والتأكيد عليها [<small>(مثال)</small>](/docs/testing-recipes.html#events).
```
Comment (Aissaoui-Ahmed) [inline content/docs/testing-environments.md]:
```suggestion
تمامًا كما في المتصفح الحقيقي ، تتيح لنا jsdom تصميم تفاعلات المستخدم ؛ يمكن للاختبارات إرسال الأحداث في عقدات DOM، ثم مراقبة الآثار الجانبية لهذه الإجراءات والتأكيد عليها [<small>(مثال)</small>](/docs/testing-recipes.html#events).
```

File: content/docs/testing-environments.md [hunk 15]
```diff
@@ -1,58 +1,59 @@
...
+تحتوي بعض المكونات على تبعيات للوحدات النمطية التي قد لا تعمل بشكل جيد في بيئات الاختبار ، أو ليست ضرورية لاختباراتنا. قد يكون من المفيد الاستغناء عن هذه الوحدات بشكل انتقائي مع بدائل مناسبة [<small>(مثال)</small>](/docs/testing-recipes.html#mocking-modules).
 
-### Mocking timers {#mocking-timers}
+على Node.js ، يقوم منفذى الاختبارات مثل Jest [بدعم محاكاة الوحدات](https://jestjs.io/docs/en/manual-mocks). يمكنك أيضًا استخدام مكتبات مثل [`mock-require`](https://www.npmjs.com/package/mock-require).
 
-Components might be using time-based functions like `setTimeout`, `setInterval`, or `Date.now`. In testing environments, it can be helpful to mock these functions out with replacements that let you manually "advance" time. This is great for making sure your tests run fast! Tests that are dependent on timers would still resolve in order, but quicker [<small>(example)</small>](/docs/testing-recipes.html#timers). Most frameworks, including [Jest](https://jestjs.io/docs/en/timer-mocks), [sinon](https://sinonjs.org/releases/v7.3.2/fake-timers/) and [lolex](https://github.com/sinonjs/lolex), let you mock timers in your tests.
+### محاكاة العداد {#mocking-timers}
```
Comment (Aissaoui-Ahmed) [inline content/docs/testing-environments.md]:
```suggestion
### محاكاة المؤقتات {#mocking-timers}
```

**Comment by mohamedsgap** [comment]
@Aissaoui-Ahmed  @3imed-jaberi  I've modified the translation guys! check it out ;)

File: content/docs/testing-environments.md [hunk 16]
```diff
@@ -1,58 +1,57 @@
...
 
-A large portion of UI tests can be written with the above setup: using Jest as a test runner, rendered to jsdom, with user interactions specified as sequences of browser events, powered by the `act()` helper [<small>(example)</small>](/docs/testing-recipes.html). For example, a lot of React's own tests are written with this combination.
+إذا كنت تكتب مكتبة تختبر في الغالب السلوك الخاص بالمتصفح ، وتتطلب سلوك المتصفح الأصلي مثل الـ layout أو حقول الإدخال الحقيقية ، يمكنك استخدام إطار عمل مثل [mocha.](https://mochajs.org/)
 
-If you're writing a library that tests mostly browser-specific behavior, and requires native browser behavior like layout or real inputs, you could use a framework like [mocha.](https://mochajs.org/)
+في بيئة لا يمكنك فيها محاكاة DOM (على سبيل المثال ، اختبار مكونات React Native على Node.js) ، يمكنك استخدام [مساعدي محاكاة الأحداث](https://reactjs.org/docs/test-utils.html#simulate) لمحاكاة التفاعلات مع العناصر. بالتناوب ، يمكنك استخدام `fireEvent` المساعد من <span dir='ltr'>[`@ testing-library / react-native`](https://testing-library.com/docs/native-testing-library)<span>.
```
Comment (Aissaoui-Ahmed) [inline content/docs/testing-environments.md]:
```suggestion
في بيئة لا يمكنك فيها محاكاة DOM (على سبيل المثال ، اختبار مكونات React Native على Node.js) ، يمكنك استخدام [مساعدي محاكاة الأحداث](https://reactjs.org/docs/test-utils.html#simulate) لمحاكاة التفاعلات مع العناصر. بالتناوب ، يمكنك استخدام `fireEvent` المساعد من <span dir='ltr'>[`@ testing-library/react-native`](https://testing-library.com/docs/native-testing-library)<span>.
```

File: content/docs/testing-environments.md [hunk 17]
```diff
@@ -1,58 +1,57 @@
...
 
-In an environment where you _can't_ simulate a DOM (e.g. testing React Native components on Node.js), you could use [event simulation helpers](https://reactjs.org/docs/test-utils.html#simulate) to simulate interactions with elements. Alternately, you could use the `fireEvent` helper from [`@testing-library/react-native`](https://testing-library.com/docs/native-testing-library).
+اطارات عمل مثل [Cypress](https://www.cypress.io/) و [puppeteer](https://github.com/GoogleChrome/puppeteer) و [webdriver](https://www.seleniumhq.org/projects / webdriver /) مفيدة لتشغيل [اختيبارات end-to-end](#end-to-end-tests-aka-e2e-tests).
```
Comment (Aissaoui-Ahmed) [inline content/docs/testing-environments.md]:
```suggestion
اطارات عمل مثل [Cypress](https://www.cypress.io/) و [puppeteer](https://github.com/GoogleChrome/puppeteer) و [webdriver](https://www.seleniumhq.org/projects/webdriver /) مفيدة لتشغيل [اختيبارات end-to-end](#end-to-end-tests-aka-e2e-tests).
```

File: content/docs/testing-environments.md [hunk 18]
```diff
@@ -1,58 +1,57 @@
...
 
-Frameworks like [Cypress](https://www.cypress.io/), [puppeteer](https://github.com/GoogleChrome/puppeteer) and [webdriver](https://www.seleniumhq.org/projects/webdriver/) are useful for running [end-to-end tests](#end-to-end-tests-aka-e2e-tests).
+### محاكاة الدوال {#mocking-functions}
 
-### Mocking functions {#mocking-functions}
+عند كتابة الاختبارات ، نود أن نحاكى  الكود الخاص بنا الذى لا يحتوى على  مُمَاثل له فى بيئة الاختبار الخاصة بنا (على سبيل المثال التحقق من حالة `navigator.onLin` داخل Node.js). يمكن للاختبارات أيضًا التجسس على بعض الدوال، ومراقبة كيفية تفاعل أجزاء أخرى من الاختبار معها. من المفيد عندئذ أن تكون قادرًا على محاكاة هذه الدوال باصدارات سهلة الاستخدام بشكل الانتقائى.
```
Comment (Aissaoui-Ahmed) [inline content/docs/testing-environments.md]:
```suggestion
عند كتابة الاختبارات ، نود أن نحاكى الكود الخاص بنا الذى لا يحتوى على مُمَاثل له فى بيئة الاختبار الخاصة بنا (على سبيل المثال التحقق من حالة `navigator.onLin` داخل Node.js). يمكن للاختبارات أيضًا التجسس على بعض الدوال، ومراقبة كيفية تفاعل أجزاء أخرى من الاختبار معها. من المفيد عندئذ أن تكون قادرًا على محاكاة هذه الدوال باصدارات سهلة الاستخدام بشكل الانتقائى.
```

File: content/docs/testing-environments.md [hunk 19]
```diff
@@ -1,58 +1,57 @@
...
 
-This document goes through the factors that can affect your environment and recommendations for some scenarios.
+يتناول هذا المستند العوامل التي يمكن أن تؤثر على بيئتك والتوصيات المتعلقة ببعض السيناريوهات.
 
-### Test runners {#test-runners}
+### منفذي الاختبار {#test-runners}
 
-Test runners like [Jest](https://jestjs.io/), [mocha](https://mochajs.org/), [ava](https://github.com/avajs/ava) let you write test suites as regular JavaScript, and run them as part of your development process. Additionally, test suites are run as part of continuous integration.
+- يتيح لك منفذى الاختبار مثل  [Jest](https://jestjs.io/), [mocha](https://mochajs.org/), [ava](https://github.com/avajs/ava) كتابة مجموعات اختبار على هيئه JavaScript و تشغيلها كجزء من عملية التطوير الخاصة بك. بالاضافة الى ذلك يتم تشغيل مجموعات الاختبار كجزء من التكامل المستمر (CI).
```
Comment (Aissaoui-Ahmed) [inline content/docs/testing-environments.md]:
```suggestion
- يتيح لك منفذي الاختبار مثل [Jest](https://jestjs.io/), [mocha](https://mochajs.org/), [ava](https://github.com/avajs/ava) كتابة مجموعات اختبار على هيئه JavaScript و تشغيلها كجزء من عملية التطوير الخاصة بك. بالاضافة الى ذلك يتم تشغيل مجموعات الاختبار كجزء من التكامل المستمر (CI).
```

File: content/docs/testing-environments.md [hunk 20]
```diff
@@ -1,58 +1,57 @@
...
+- Jest متوافق على نطاق واسع مع مشاريع React, و دعم مميزات جديده مثل [الوحدات النمطية ](#moking-modules)و [العداد](#moking-timers) و دعم [`jsdom`](#mocking-a-rendering-surface`). **اذا كنت تستخدم Create React App اذن [Jest موجود بالفعل](https://facebook.github.io/create-react-app/docs/running-tests) مع تسطيب افتراضي مفيد**.
+- المكتبات مثل [mocha](https://mochajs.org/#running-mocha-in-the-browser) تعمل بشكل جيد في بيئات المتصفح الحقيقي ، ويمكن أن تساعد في الاختبارات التي تحتاجها بشكل صريح.
```
Comment (Aissaoui-Ahmed) [inline content/docs/testing-environments.md]:
```suggestion
- المكتبات مثل [mocha](https://mochajs.org/#running-mocha-in-the-browser) تعمل بشكل جيد في بيئات المتصفح الحقيقي، ويمكن أن تساعد في الاختبارات التي تحتاجها بشكل صريح.
```

File: content/docs/testing-environments.md [hunk 21]
```diff
@@ -1,58 +1,57 @@
...
+- تُستخدم اختبارات End-to-End لاختبار تدفقات أطول عبر عدة صفحات، وتتطلب [إعدادات مختلفة](#end-to-end-tests-aka-e2e-tests).
 
-- Jest is widely compatible with React projects, supporting features like mocked [modules](#mocking-modules) and [timers](#mocking-timers), and [`jsdom`](#mocking-a-rendering-surface) support. **If you use Create React App, [Jest is already included out of the box](https://facebook.github.io/create-react-app/docs/running-tests) with useful defaults.**
-- Libraries like [mocha](https://mochajs.org/#running-mocha-in-the-browser) work well in real browser environments, and could help for tests that explicitly need it.
-- End-to-end tests are used for testing longer flows across multiple pages, and require a [different setup](#end-to-end-tests-aka-e2e-tests).
+### محاكاة واجهة التصيير {#mocking-a-rendering-surface}
 
-### Mocking a rendering surface {#mocking-a-rendering-surface}
+تعمل الاختبارات غالبًا في بيئة دون الوصول إلى واجهة التصيير الحقيقية مثل المتصفح. بالنسبة لهذه البيئات ، نوصي بمحاكاة متصفح باستخدام [`jsdom`](https://github.com/jsdom/jsdom) ، وهو تطبيق متصفح خفيف الوزن يعمل داخل Node.js.
```
Comment (Aissaoui-Ahmed) [inline content/docs/testing-environments.md]:
```suggestion
تعمل الاختبارات غالبًا في بيئة دون الوصول إلى واجهة التصيير الحقيقية مثل المتصفح. بالنسبة لهذه البيئات، نوصي بمحاكاة متصفح باستخدام [`jsdom`](https://github.com/jsdom/jsdom) ، وهو تطبيق متصفح خفيف الوزن يعمل داخل Node.js.
```

File: content/docs/testing-environments.md [hunk 22]
```diff
@@ -1,58 +1,57 @@
...
 
-Components might be using time-based functions like `setTimeout`, `setInterval`, or `Date.now`. In testing environments, it can be helpful to mock these functions out with replacements that let you manually "advance" time. This is great for making sure your tests run fast! Tests that are dependent on timers would still resolve in order, but quicker [<small>(example)</small>](/docs/testing-recipes.html#timers). Most frameworks, including [Jest](https://jestjs.io/docs/en/timer-mocks), [sinon](https://sinonjs.org/releases/v7.3.2/fake-timers/) and [lolex](https://github.com/sinonjs/lolex), let you mock timers in your tests.
+في بعض الأحيان ، قد لا ترغب في محاكاة العداد. على سبيل المثال ، ربما تقوم باختبار رسم متحرك ، أو تتفاعل مع نقطة نهاية حساسة للتوقيت (مثل واجهة برمجة التطبيقات من نوع API rate limiter). تتيح لك المكتبات التي بها محاكاه العداد تمكينها وتعطيلها على أساس كل اختبار / مجموعة ، بحيث يمكنك اختيار كيفية تشغيل هذه الاختبارات بشكل صريح.
```
Comment (Aissaoui-Ahmed) [inline content/docs/testing-environments.md]:
```suggestion
في بعض الأحيان ، قد لا ترغب في محاكاة العداد. على سبيل المثال ، ربما تقوم باختبار رسم متحرك ، أو تتفاعل مع نقطة نهاية حساسة للتوقيت (مثل واجهة برمجة التطبيقات من نوع API rate limiter). تتيح لك المكتبات التي بها محاكاه العداد تمكينها وتعطيلها على أساس كل اختبار/مجموعة ، بحيث يمكنك اختيار كيفية تشغيل هذه الاختبارات بشكل صريح.
```

**Comment by mohamedsgap** [comment]
@Aissaoui-Ahmed  @3imed-jaberi  I've made the last changed, check it out!!

File: content/docs/testing-environments.md [hunk 23]
```diff
@@ -1,58 +1,60 @@
...
+### منفذي الاختبار {#test-runners}
 
-- Jest is widely compatible with React projects, supporting features like mocked [modules](#mocking-modules) and [timers](#mocking-timers), and [`jsdom`](#mocking-a-rendering-surface) support. **If you use Create React App, [Jest is already included out of the box](https://facebook.github.io/create-react-app/docs/running-tests) with useful defaults.**
-- Libraries like [mocha](https://mochajs.org/#running-mocha-in-the-browser) work well in real browser environments, and could help for tests that explicitly need it.
-- End-to-end tests are used for testing longer flows across multiple pages, and require a [different setup](#end-to-end-tests-aka-e2e-tests).
+يتيح لك منفذى الاختبار مثل  [Jest](https://jestjs.io/), [mocha](https://mochajs.org/), [ava](https://github.com/avajs/ava) كتابة مجموعات اختبار على هيئه JavaScript و تشغيلها كجزء من عملية التطوير الخاصة بك. بالاضافة الى ذلك يتم تشغبل مجموعات الاختبار كجزء من التكامل المستمر.
```
[the remark this answers is not in the document]
Comment (atefBB) [reply content/docs/testing-environments.md]: <div dir="rtl">"يتيح لك منفذى الاختبار" الصواب يتيح لك منفذو الاختبار.</div>

[the remark this answers is not in the document]
Comment (atefBB) [reply content/docs/testing-environments.md]: <div dir="rtl">"يتيح لك منفذى الاختبار" الصواب يتيح لك منفذو الاختبار.</div>

**Pull request merged**